sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java

```diff
@@ -0,0 +1,284 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.RunnableOnService;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.testing.TestStream;
+import org.apache.beam.sdk.testing.UsesTestStream;
+import org.apache.beam.sdk.testing.UsesTimersInParDo;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.hamcrest.Matchers;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.BeforeClass;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.LoggerFactory;
+
+/** Test Class for {@link BatchingParDo}. */
+@RunWith(JUnit4.class)
+public class BatchingParDoTest implements Serializable {
+  private static final int BATCH_SIZE = 3;
+  private static final long NUM_ELEMENTS = 100;
+  private static final int ALLOWED_LATENESS = 0;
+  private static final int TIMESTAMP_INTERVAL = 1;
+  private static final long WINDOW_DURATION = 5;
+  private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(BatchingParDoTest.class);
+  private static ArrayList<TimestampedValue<KV<String, String>>> data;
+  private static SimpleFunction<Iterable<String>, Iterable<String>> perBatchFn;
+  private static BatchingParDo<String, String, String> batchingParDo;
+  private static Instant startInstant;
+  @Rule public transient TestPipeline pipeline = TestPipeline.create();
+
+  @BeforeClass
+  public static void initialize() {
+    startInstant = new Instant(0);
+    data = createTestData();
+    perBatchFn =
+        new SimpleFunction<Iterable<String>, Iterable<String>>() {
+          @Override
+          public Iterable<String> apply(Iterable<String> input) {
+            ArrayList<String> output = new ArrayList<>();
+            for (String element : input) {
+              output.add(element + "2");
+            }
+            return output;
+          }
+        };
+    batchingParDo = BatchingParDo.via(BATCH_SIZE, perBatchFn);
```

Comment:
You mean that the gain of readability is worth creating ParDo twice? Well in a test, I guess creating twice the ParDo is ok